aconfig/aconfig.py

```diff
@@ -83,7 +86,35 @@ def __deepcopy__(self, memo):
         '''This enables deepcopy to successfully copy a Config object, despite
         the default value semantics
         '''
-        return AttributeAccessDict(copy.deepcopy(dict(self)))
+        return self.__class__(copy.deepcopy(dict(self)))
+
+
+class ImmutableAttributeAccessDict(AttributeAccessDict):
+    """This class subclasses AttributeAccessDict and removes the setters,
+    to allow the creation of immutable dicts.
+
+    Using inheritance this way allows the dicts to be recursively created via
+    AttributeAccessDict, while maintaining nested immutability.
+    """
+
+    def __init__(self, input_map, *_):
+        """See :func:`~aconfig.aconfig.AttributeAccessDict.__init__`"""
+        if not isinstance(input_map, dict):
+            raise TypeError('`input_map` argument should be of type dict, but found type: <{0}>'.format(
+                type(input_map)))
+        # 🌶️🌶️🌶️: we explicitly cast back down to `dict` for the immutable case
+        # If we were to build an immutable dict from the top-down, that would
+        # obviously fail.
+        input_map = dict(input_map)
+        # Invoke the AttributeAccessDict initializer
+        super().__init__(input_map)
+
+    def __setitem__(self, key, value):
+        raise AttributeError("ImmutableAttributeAccessDict does not support attribute assignment")
```

Comment:
Heh, I actually think this _should_ be a `TypeError` to keep with what happens if you try to use item assignment on a `mappingproxy`:

```py
Foo.__dict__["asdf"] = 1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'mappingproxy' object does not support item assignment
```

Reply:
ah fair